### PR TITLE
refactor(graph-proxy): update science-group filter semantics

### PR DESCRIPTION
**Pull request by MaxHerbs**

### CODE REVIEW

File: backend/graph-proxy/src/graphql/filters.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        self.science_group.generate_labels(&mut label_selectors);
+
+        label_selectors.join(",")
+    }
+}
+
+/// Supported DLS science groups
+#[derive(InputObject, Debug, Default, Clone)]
+pub struct ScienceGroup {
```
[the remark this answers is not in the document]
Comment (davehadley) [reply backend/graph-proxy/src/graphql/filters.rs:50]: I would have made the available ScienceGroup's an enum.

Then I would have called this struct ScienceGroupFilter that contains a HashMap<ScienceGroup, bool> or something similar.

Then future changes to the ScienceGroup is a one line change to the enum.

File: backend/graph-proxy/src/graphql/filters.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    /// Examples Group
+    examples: Option<bool>,
+}
+
+impl GraphFilter for ScienceGroup {
+    fn generate_labels(&self, labels: &mut Vec<String>) {
+        let label_prefix = "workflows.diamond.ac.uk/science-group-";
+
+        let fields = [
+            ("mx", self.mx),
```
Comment (davehadley) [inline backend/graph-proxy/src/graphql/filters.rs:66]: if you choose to go down the enum route, these group name strings could be part of the enum

File: backend/graph-proxy/src/graphql/filters.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
+use async_graphql::InputObject;
+use url::Url;
+
+/// Build labels to apply query to workflows API
+pub trait GraphFilter {
```
Comment (davehadley) [inline backend/graph-proxy/src/graphql/filters.rs:5]: I think this trait can be private for now until it's needed in other modules.

File: backend/graph-proxy/src/graphql/filters.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    /// The science group owning the template eg imaging
+    science_group: Option<ScienceGroup>,
+}
+
+impl WorkflowTemplatesFilter {
+    /// Generates and applies all the filters
+    pub fn generate_filters(&self, url: &mut Url) {
+        let labels = &self.create_label_selection();
+        url.query_pairs_mut()
+            .append_pair("listOptions.labelSelector", labels);
```
Comment (davehadley) [inline backend/graph-proxy/src/graphql/filters.rs:35]: does this behave correctly if labels is an empty string (i.e. no filters were added)?

File: backend/graph-proxy/src/graphql/filters.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::graphql::filters::{ScienceGroup, WorkflowTemplatesFilter};
+
+    #[tokio::test]
+    async fn science_group_filter() {
+        let science_group = ScienceGroup {
```
Comment (davehadley) [inline backend/graph-proxy/src/graphql/filters.rs:86]: If possible, consider splitting into three separate tests. This makes debugging easier as you get the result from all 3 asserts at once.

File: backend/graph-proxy/src/graphql/filters.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            magnetic_materials: None,
+        };
+        let filters = WorkflowTemplatesFilter {
+            science_group: Some(science_group),
+        };
+        let label_selectors = filters.create_label_selection();
+        assert_eq!(
+            label_selectors,
+            "workflows.diamond.ac.uk/science-group-examples=true"
+        );
```
Comment (davehadley) [inline backend/graph-proxy/src/graphql/filters.rs:126]: Consider adding a test for the ScienceGroup{None, None, None, None} case.

**Review by davehadley — CHANGES_REQUESTED** [review]
Thanks for putting this together.

Moving the filter logic into its own module makes sense and makes the codebase easier to maintain and extend.

One thing to keep in mind for future pull requests: this one mixes a feature that changes the public interface (the accepted format for specifying science-group in templates) with some refactoring. It helps the review if these can be two separate commits (so I can just focus on the important changes which is the feature change).

For release-please I think this commit message should be changed to feat.

File: backend/graph-proxy/src/graphql/filters.rs (hunk 4 above)
Comment (MaxHerbs) [reply backend/graph-proxy/src/graphql/filters.rs:35]: if no filters are added then the generate_filters function should never be called to start with so that should be fine
